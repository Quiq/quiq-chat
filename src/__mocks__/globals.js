let burned = false;
export const getBurned = () => burned;
export const setBurned = (shouldBurn?: boolean) => {
  burned = typeof shouldBurn === 'boolean' ? shouldBurn : true;
};
export const setGlobals = jest.fn();
export const checkRequiredSettings = jest.fn();
export const getHost = () => 'someHost';
export const getContactPoint = () => 'someContactPoint';
export const getPublicApiUrl = () => 'somePublicApiUrl';
export const getUrlForContactPoint = () => 'someUrlForContactPoint';
export const isActive = jest.fn().mockReturnValue(true);
