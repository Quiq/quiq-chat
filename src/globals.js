// @flow
import Raven from 'raven-js';
import {getTenantFromHostname} from 'Utils/utils';
import type {QuiqChatSettings} from 'types';

let burned = false;
export const getBurned = () => burned;
export const setBurned = () => {
  burned = true;
};

let quiqChatSettings: QuiqChatSettings;

const defaults = {
  CONTACT_POINT: 'default',
};

export const setGlobals = (globals: QuiqChatSettings) => {
  quiqChatSettings = Object.assign({}, defaults, globals);

  // Update Raven if host is defined
  if (quiqChatSettings.HOST) {
    Raven.setTagsContext({
      tenant: getTenantFromHostname(quiqChatSettings.HOST),
    });
  }
};

export const checkRequiredSettings = () => {
  if (getBurned()) {
    throw new Error('Client in bad state. Aborting call.');
  }

  if (!quiqChatSettings || !quiqChatSettings.HOST || !quiqChatSettings.CONTACT_POINT) {
    throw new Error(
      `
      HOST and CONTACT_POINT must be configured to call Quiq Messaging.
      Did you forget to call init?
      `,
    );
  }
};

export const getContactPoint = () => quiqChatSettings && quiqChatSettings.CONTACT_POINT;

const getHost = (cached: boolean) => {
  if (
    quiqChatSettings.HOST.includes('goquiq.com') ||
    quiqChatSettings.HOST.includes('quiq.dev') ||
    (quiqChatSettings.HOST.includes('quiq-api') && cached)
  ) {
    return quiqChatSettings.HOST;
  }

  try {
    const vanityName = quiqChatSettings.HOST.split('.')[0].split('https://')[1];
    return `https://${vanityName}.goquiq.com`;
  } catch (e) {
    return quiqChatSettings.HOST;
  }
};

export const getPublicApiUrl = (cached?: boolean = false) => `${getHost(cached)}/api/v1/messaging`;

export const getUrlForContactPoint = (cached?: boolean = false) =>
  `${getHost(cached)}/api/v1/messaging/chat/${quiqChatSettings.CONTACT_POINT}`;

export const getSessionApiUrl = (cached?: boolean = false) => `${getHost(cached)}/session/web`;

export const getGenerateUrl = (cached?: boolean = false) => {
  return `${getHost(cached)}/api/v1/token/generate`;
};
