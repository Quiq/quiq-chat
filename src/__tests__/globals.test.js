// @flow

jest.mock('../logging');
jest.mock('../Utils/utils.js');

import * as Globals from '../globals';

describe('Globals module', () => {
  describe('before globals have been set', () => {
    describe('checkRequiredSettings', () => {
      it('throws an error', () => {
        expect(Globals.checkRequiredSettings).toThrowError();
      });
    });
  });

  describe('after globals have been set', () => {
    describe('checkRequiredSettings', () => {
      it('does not throw an error', () => {
        Globals.setGlobals({HOST: 'https://testhost.goquiq.com', CONTACT_POINT: 'test'});
        expect(Globals.checkRequiredSettings).not.toThrowError();
      });
    });

    describe('getContactPoint', () => {
      it('returns the contact point', () => {
        expect(Globals.getContactPoint()).toBe('test');
      });
    });

    describe('getPublicApiUrl', () => {
      it('returns the base url for the public api', () => {
        expect(Globals.getPublicApiUrl()).toMatchSnapshot();
      });
    });

    describe('getUrlForContactPoint', () => {
      it('returns the url for the api for the current contact point', () => {
        expect(Globals.getUrlForContactPoint()).toMatchSnapshot();
      });
    });
  });
});
