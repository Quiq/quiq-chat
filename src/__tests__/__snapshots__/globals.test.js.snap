// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Globals module after globals have been set getPublicApiUrl returns the base url for the public api 1`] = `"https://testhost.goquiq.com/api/v1/messaging"`;

exports[`Globals module after globals have been set getUrlForContactPoint returns the url for the api for the current contact point 1`] = `"https://testhost.goquiq.com/api/v1/messaging/chat/test"`;
